src/ImageStack.cpp

```diff
@@ -50,75 +51,104 @@ int ImageStack::addImage(Image && i) {
-                    histograms[c][i] += histogramsThr[c][i];
+            for (int c_channel = 0; c_channel < 4; ++c_channel) {
+                std::vector<size_t>::size_type histogram_size, luminance_val;
+                histogram_size = histograms[c_channel].size();
+                for (luminance_val = 0; luminance_val < histogram_size; ++luminance_val) {
+                    histograms[c_channel][luminance_val] += histograms_thr[c_channel][luminance_val];
                 }
             }
         }
     }
 
-    const size_t threshold = width * height / 10000;
+    // used to ignore luminances with frequentcy less htan 0.0001
+    const size_t occurance_threshold = width * height / 10000;
 
-    uint16_t maxPerColor[4] = {0, 0, 0, 0};
+    // maximum "significant" luminance per color channel
+    uint16_t max_lum_per_color[4] = {0, 0, 0, 0};
 
-    for (int c = 0; c < 4; ++c) {
-        for (int i = histograms[c].size() - 1; i >= 0; --i) {
-            const size_t v = histograms[c][i];
-            if (v > threshold) {
-                maxPerColor[c] = i;
+    for (int c_channel = 0; c_channel < 4; ++c_channel) {
+        // start from the highest value in the histograms (brightest)
+        for (int luminance_val = histograms[c_channel].size() - 1; luminance_val >= 0; --luminance_val) {
+            const size_t frequency = histograms[c_channel][luminance_val];
+            // ignore if it has a low occurance (frequency) in the image
+            if (frequency > occurance_threshold) {
+                max_lum_per_color[c_channel] = luminance_val;
                 break;
             }
         }
     }
 
 
-    uint16_t maxPerColors = std::max(maxPerColor[0], std::max(maxPerColor[1],std::max(maxPerColor[2], maxPerColor[3])));
-    satThreshold = params.max == 0 ? maxPerColors : params.max;
+    uint16_t max_luminance = std::max(max_lum_per_color[0], std::max(max_lum_per_color[1],std::max(
+        max_lum_per_color[2], max_lum_per_color[3])));
+    saturation_threshold = params.max == 0 ? max_luminance : params.max;
 
-    if(maxPerColors > 0) {
-        satThreshold = std::min(satThreshold, maxPerColors);
+    if(max_luminance > 0) {
+        saturation_threshold = std::min(saturation_threshold, max_luminance);
     }
 
-    if (!useCustomWl) { // only scale when no custom white level was specified
-        satThreshold *= 0.99;
+    // only scale when no custom white level was specified
+    if (!use_custom_white_level) {
+        // The saturation threshold needs to go a little bit "before" the
+        // highest notable luminance so that it is considered oversaturated
+        saturation_threshold *= 0.99;
     }
 
-    Log::debug( "Using white level ", satThreshold );
+    Log::debug( "Using white level ", saturation_threshold );
 
-    for (auto& i : images) {
-        i.setSaturationThreshold(satThreshold);
+    for (auto& img : images) {
```

Comment:
If it doesn't clash with another name, I'd even go for `image`. It's always good to have the singular of a plural form in range based for loops, and clang-tidy will even choose such a name when [converting](https://clang.llvm.org/extra/clang-tidy/checks/modernize-loop-convert.html).